src/common/ad_mem.c

```diff
@@ -3,5 +3,1917 @@
+	ad = (struct ad_arena_df *)((unsigned long)ptr & ~ARENA_SIZE_MASK);
+	return ad->ad_addr + off;
+}
+
+daos_off_t
+ad_ptr2addr(struct ad_blob_handle bh, void *ptr)
+{
+	return blob_ptr2addr(bh.bh_blob, ptr);
+}
+
+static int
+group_size_cmp(const void *p1, const void *p2)
+{
+	const struct ad_group_df *gd1 = p1;
+	const struct ad_group_df *gd2 = p2;
+	const struct ad_group	 *gp1 = group_df2ptr(gd1);
+	const struct ad_group	 *gp2 = group_df2ptr(gd2);
+	int			  f1;
+	int			  f2;
+
+	if (gd1->gd_unit < gd2->gd_unit)
+		return -1;
+	if (gd1->gd_unit > gd2->gd_unit)
+		return 1;
+
+	f1 = gd1->gd_unit_free - (gp1 ? gp1->gp_unit_rsv : 0);
+	f2 = gd2->gd_unit_free - (gp2 ? gp2->gp_unit_rsv : 0);
+	if (f1 < f2)
+		return -1;
+	if (f1 > f2)
+		return 1;
+
+	/* use address to identify a specified group */
+	if (gd1->gd_addr < gd2->gd_addr)
+		return -1;
+	if (gd1->gd_addr > gd2->gd_addr)
+		return 1;
+	return 0;
+}
+
+static int
+group_addr_cmp(const void *p1, const void *p2)
+{
+	const struct ad_group_df *gd1 = p1;
+	const struct ad_group_df *gd2 = p2;
+
+	if (gd1->gd_addr < gd2->gd_addr)
+		return -1;
+	if (gd1->gd_addr > gd2->gd_addr)
+		return 1;
+
+	/* two groups have the same address? */
+	D_ASSERT(0);
+	return 0;
+}
+
+static int
+arena_locate(struct ad_blob *blob, bool create, uint32_t arena_id, struct ad_arena_df **ad_p)
+{
+	struct ad_blob_df  *bd = blob->bb_df;
+	bool		    created;
+	bool		    reserved;
+
+	if ((arena_id << ARENA_SIZE_BITS) > blob_size(blob))
+		return -DER_INVAL;
+
+	created = isset((uint8_t *)bd->bd_bmap, arena_id);
+	reserved = isset((uint8_t *)blob->bb_bmap_rsv, arena_id);
+	if (create == (created || reserved))
+		return create ? -DER_EXIST : -DER_NONEXIST;
+
+	/* Arena is the header of each page */
+	*ad_p = (struct ad_arena_df *)blob->bb_pages[arena_id].pa_rpg;
+	return 0;
+}
+
+int
+arena_load(struct ad_blob *blob, uint32_t arena_id, struct ad_arena **arena_p)
+{
+	struct ad_arena	    *arena = NULL;
+	struct ad_arena_df  *ad = NULL;
+	struct ad_group_df  *gd;
+	int		     i;
+	int		     rc;
+
+	rc = arena_locate(blob, false, arena_id, &ad);
+	if (rc) {
+		D_ERROR("No available arena\n");
+		return rc;
+	}
+
+	if (ad->ad_magic != ARENA_MAGIC) {
+		D_ERROR("Invalid arena magic: %x/%x\n", ad->ad_magic, ARENA_MAGIC);
+		return -DER_PROTO;
+	}
+
+	if (ad->ad_incarnation != blob_incarnation(blob)) {
+		ad->ad_incarnation = blob_incarnation(blob);
+		ad->ad_back_ptr = 0; /* clear stale back-pointer to DRAM */
+	}
+
+	if (ad->ad_back_ptr) {
+		arena = arena_df2ptr(ad);
+		D_ASSERT(arena->ar_df == ad);
+		arena->ar_ref++;
+		if (arena->ar_ref == 1) { /* remove from LRU */
+			d_list_del_init(&arena->ar_link);
+			blob->bb_ars_lru_size--;
+		}
+		goto out;
+	}
+	/* no cached arena, allocate it now */
+
+	arena = alloc_arena(blob, false);
+	if (!arena)
+		return -DER_NOMEM;
+
+	/* NB: don't need to write to WAL,
+	 * NB: stale pointer can be detected by incarnation.
+	 */
+	ad->ad_back_ptr = (unsigned long)arena;
```

Comment:
Do we really need to put a DRAM pointer into durable format?

Reply:
as I explained, it can avoid another index, which is good for performance.